pkg/generator/gen.go

```diff
@@ -0,0 +1,340 @@
+		// arr[0] -> group, arr[1] -> version
+		arr := strings.Split(gv, ":")
+		if len(arr) != 2 {
+			return fmt.Errorf("input to --group-version must be in <group>:<version> format, ex: rbac:v1. Got %q", gv)
+		}
+
+		// input path is converted to <inputDir>/pkg/apis/<group>/<version>.
+		// example for input directory of "k8s.io/client-go/kubernetes", it would
+		// be converted to "k8s.io/client-go/kubernetes/pkg/apis/rbac/v1".
+		input := filepath.Join(f.InputDir, "pkg", "apis", arr[0], arr[1])
+		groups := []types.GroupVersions{}
+		builder := args.NewGroupVersionsBuilder(&groups)
+		_ = args.NewGVPackagesValue(builder, []string{input})
+
+		g.groupVersions = append(g.groupVersions, groups...)
+	}
+	return nil
+}
+
+// generate first genrates the wrapper for all the interfaces provided in the input.
+// Then for each type defined in the input, it recursively wraps the subsequent
+// interfaces to be kcp-aware.
+func (g *Generator) generate(ctx *genall.GenerationContext) error {
+	err := g.writeWrappedClientSet()
+	if err != nil {
+		return err
+	}
+	return g.generateSubInterfaces(ctx)
+}
+
+func (g *Generator) writeWrappedClientSet() error {
+	out := new(bytes.Buffer)
+	if err := wrtieHeader(out); err != nil {
+		return err
+	}
+
+	wrappedInf, err := internal.NewInterfaceWrapper(g.interfaceName, g.inputDir, g.groupVersions, out)
+	if err != nil {
+		return err
+	}
+
+	if err := wrappedInf.WriteContent(); err != nil {
+		return err
+	}
+
+	outBytes := out.Bytes()
+	formattedBytes, err := format.Source(outBytes)
+	if err != nil {
+		return err
+	} else {
+		outBytes = formattedBytes
+	}
+	return g.writeContent(outBytes, clientSetFilename)
+}
+
+// wrtieContents creates a new file under the path <outputDir>/generated with
+// the specified filename and write contents to it.
+func (g *Generator) writeContent(outBytes []byte, filename string) error {
+	path := filepath.Join(g.outputDir, generatedFolderName)
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		err = os.MkdirAll(path, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	outputFile, err := os.Create(filepath.Join(path, filename))
+	if err != nil {
+		return err
+	}
+	defer outputFile.Close()
+
+	n, err := outputFile.Write(outBytes)
+	if err != nil {
+		return err
+	}
+	if n < len(outBytes) {
+		return err
+	}
+	return nil
+}
+
+func wrtieHeader(out io.Writer) error {
+	n, err := out.Write([]byte(internal.HeaderText))
+	if err != nil {
+		return err
+	}
+
+	if n < len([]byte(internal.HeaderText)) {
+		return errors.New("header text was not written properly.")
+	}
+	return nil
+}
+
+func (g *Generator) generateSubInterfaces(ctx *genall.GenerationContext) error {
+	for _, gv := range g.groupVersions {
+		// the group version flag from input is allowed to have only one version
+		// for now.
+		// TODO: modify this to accept a list instead
+		version := gv.Versions[0]
+		path := filepath.Join(g.inputDir, "pkg", "apis", gv.Group.String(), string(version.Version))
```

Comment:
This puts everything for a GV into a single file. Would you want to use a directory hierarchy like client-go's client-gen does?

Reply:
Modified this to be:
- clientset wrappers would be inside `<outputDir>/generated/clientset.go`.
- individual typed client wrappers would be inside `<outputDir>/generated/${GROUP}/${VERSION}/${group_version}.go`.

Or do we want it to be the exact same [hierarchy](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/generating-clientset.md#output-of-client-gen) as code-gen? The reason I didn't make it the same was in case the user wants to have the output generated in the same folder as code-gen scaffolds are present, then we would end up overwriting (unless we ensure the filename to be always different).

Reply:
I don't think we'd want to encourage (or support) generating our wrapper clients into the same package as Kube's client-gen. I think matching the hierarchy in e.g. https://github.com/kcp-dev/kcp/tree/main/pkg/client/clientset/versioned would be nice, if it's not too much work.

Reply:
Modified to accept the `clientset-name` flag and match the same hierarchy.